docs/source/format/DissociatedIPC.rst

```diff
@@ -0,0 +1,333 @@
+
+For each IPC message in the stream of data, a **tagged** message **MUST** be sent on the data
+stream if that message has a body (i.e. a Record Batch or Dictionary message). The 
+:term:`tag <Tag>` for each message should be structured as follows:
+
+* The *least significant* 4-bytes (bits 0 - 31) of the tag should be the 32-bit, little-endian sequence 
+number of the message.
+* The *most significant* byte (bits 56 - 63) of the tag indicates the message body **type** as an 8-bit
+unsigned integer. Currently only two message types are specified, but more can be added as
+needed to expand the protocol:
+  0) The body contains the raw body buffer bytes as a packed buffer (i.e. the standard IPC
+  format body bytes)
+  1) The body contains a series of unsigned, little-endian 64-bit integer pairs to represent
+  either shared or remote memory, schematically structured as
+    - The first two integers (e.g. the first 16 bytes) represent the *total* size (in bytes)
+    of all buffers and the number of buffers in this message (and thus the number of following
+    pairs of ``uint64``)
+    - Each subsequent pair of ``uint64`` values are an address/offset followed the length of
+    that particular buffer.
+* All unspecified bits (bits 32 - 55) of the tag are *reserved* for future use by potential updates
+to this protocol. For now they **MUST** be 0.
+
+.. note::
+
+  Any shared/remote memory addresses that are sent across **MUST** be kept alive by the server
+  until a corresponding tagged ``<free_data>`` message is received. If the client disconnects
+  before sending any ``<free_data>`` messages, it can be assumed to be safe to clean up the memory
+  if desired by the server.
+
+After sending the last tagged IPC body message, the server should maintain the connection and wait
+for tagged ``<free_data>`` messages. The structure of these ``<free_data>`` messages is simple:
+1 or more unsigned, little-endian 64-bit integers which indicate the addresses/offsets that can
+be freed. 
+
+Once there are no more outstanding addresses to be freed, the work for this connection is complete.
+
+Client Sequence
+---------------
+
+A client for this protocol needs to asynchronously handle both the data and metadata streams of
+messages which may either both come from the same server or different servers. Below is a flowchart
+showing how a client might handle the metadata and data streams:
+
+.. figure:: ./DissociatedIPC/ClientFlowchart.mmd.svg
+
+#. First the client sends a tagged message using the ``<want_data>`` value it was provided in the
+URI as the tag, and the opaque ID as the body.
+  * If the metadata and data servers are separate, then a ``<want_data>`` message needs to be sent
+  separately to each. 
+  * In either scenario, the metadata and data streams can be processed concurrently and/or asynchronously
+  depending on the nature of the transports.
+#. For each **untagged** message the client receives in the metadata stream:
```

Comment:
How does this actually work in the stated motivation of being able to use Flight RPC for metadata only? gRPC only supports delimited messages, not tagged messages. You'd need some way to initiate a gRPC request alongside a UCX connection, or something.

Reply:
That would follow the pattern of separate transports / streams for the data and metadata, which would have had two URIs (one for each). The gRPC delimited, untagged messages would be the metadata messages as described, while UCX would send the tagged data messages.